python/pyspark/sql/connect/client/reattach.py

```diff
@@ -58,28 +58,38 @@ class ExecutePlanResponseReattachableIterator(Generator):
 
     # Lock to manage the pool
     _lock: ClassVar[RLock] = RLock()
-    _release_thread_pool: Optional[ThreadPool] = ThreadPool(os.cpu_count() if os.cpu_count() else 8)
+    _release_thread_pool_instance = None
+
+    @classmethod  # type: ignore[misc]
+    @property
+    def _release_thread_pool(cls) -> ThreadPool:
+        with cls._lock:
+            if cls._release_thread_pool_instance is None:
+                cls._release_thread_pool_instance = ThreadPool(
+                    os.cpu_count() if os.cpu_count() else 8
+                )
+            return cls._release_thread_pool_instance
```

Comment:
Nit: The goal of this code is to initialize the code exactly once. The above implementation requires all callers to always acquire the lock even if it is not needed. For that reason and to avoid congestion on the lock, we should do a first check outside the critical path:

```suggestion
        # Perform a first check outside the critical path.
        if not cls._release_thread_pool_instance is None:
          return cls._release_thread_pool_instance
        with cls._lock:
            if cls._release_thread_pool_instance is None:
                cls._release_thread_pool_instance = ThreadPool(
                    os.cpu_count() if os.cpu_count() else 8
                )
            return cls._release_thread_pool_instance
```